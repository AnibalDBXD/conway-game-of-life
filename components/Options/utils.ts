export const IMAGE_SIZE = 89;
