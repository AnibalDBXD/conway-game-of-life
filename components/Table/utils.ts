import { IChangeCells, ICells } from './types';

export const createCell = ({ X, Y, Cells }: IChangeCells): ICells => {
  const newCells = [...Cells];
  newCells[Y][X] = true;
  return newCells;
};

export const deleteCell = ({ X, Y, Cells }: IChangeCells): ICells => {
  const newCells = [...Cells];
  newCells[Y][X] = false;
  return newCells;
};

export const getNeighbours = ({ X, Y, Cells }: IChangeCells): boolean[] => {
  const getCell = (cellX: number, cellY: number): boolean => {
    let rows: boolean[];
    if (cellY >= 0) {
      rows = Cells[cellY % Cells.length];
    } else {
      rows = Cells[Cells.length - 1];
    }
    let cell: boolean;
    if (cellX >= 0) {
      cell = rows[cellX % rows.length];
    } else {
      cell = rows[rows.length - 1];
    }
    return cell;
  };
  const upstairsNeighbor = getCell(X, Y - 1);
  const downstairsNeighnour = getCell(X, Y + 1);
  const leftNeighnour = getCell(X - 1, Y);
  const rightNeighnour = getCell(X + 1, Y);
  const upstairsNeighborAndLeft = getCell(X - 1, Y - 1);
  const upstairsNeighborAndRight = getCell(X + 1, Y - 1);
  const neighborBelowAndRight = getCell(X + 1, Y + 1);
  const neighborBelowAndLeft = getCell(X - 1, Y + 1);

  return [
    upstairsNeighbor,
    downstairsNeighnour,
    leftNeighnour,
    rightNeighnour,
    upstairsNeighborAndLeft,
    upstairsNeighborAndRight,
    neighborBelowAndRight,
    neighborBelowAndLeft,
  ];
};

export const createColumnsAndRows = (newColumnsAndRows: number): boolean[][] =>
  new Array(newColumnsAndRows)
    .fill(false)
    .map(() => new Array(newColumnsAndRows).fill(false));
