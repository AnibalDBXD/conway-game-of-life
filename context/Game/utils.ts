export enum DEFAULT_VALUES {
    TIME = 200,
    COLUMUNS_AND_ROWS = 16,
}

export enum MIN_VALUES {
    TIME = 100,
    COLUMUNS_AND_ROWS = 8
}

export enum MAX_VALUES {
    TIME = 1000,
    COLUMUNS_AND_ROWS = 40
}