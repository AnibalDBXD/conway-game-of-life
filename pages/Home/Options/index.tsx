import React, { ChangeEvent, FormEvent, useRef, useEffect } from "react";
import { IOptions } from "./types";
import { StyledForm, StyledOptions, StyledTipsAndRulesContainer } from "./styles";
import PlayPause from "./PlayPause";

const Options = ({ setPause, isPause, setNumberOfColumnsAndRows, setTime }: IOptions): JSX.Element => {
  let newTime = null;
  const MAX_TIME = 10000;
  const MIN_TIME = 100;
  const DEFAULT_TIME = 200;

  let newColumnsAndRows = null;

  const timeRef = useRef<HTMLInputElement>(null);
  const columnsAndRowsRef = useRef<HTMLInputElement>(null);

  const onChangeTime = (event: ChangeEvent<HTMLInputElement>): void => {
    event.preventDefault();
    const currentTime = Number(event.target.value);
    if (currentTime > MAX_TIME) {
      timeRef.current.value = MAX_TIME + "";
      return;
    }
    if (currentTime < MIN_TIME) {
      newTime = MIN_TIME;
      return;
    }
    newTime = (currentTime);
  };

  const onChangeColumnsAndRows = (event: ChangeEvent<HTMLInputElement>): void => {
    event.preventDefault();
    newColumnsAndRows = ((Number(event.target.value)));
  };

  const handleSubmit = (event?: FormEvent<HTMLFormElement> | FormEvent<HTMLButtonElement>): void => {
    event?.preventDefault();
    if (newTime) {
      setTime(newTime);
      timeRef.current.value = "";
    }
    if (newColumnsAndRows) {
      setNumberOfColumnsAndRows(newColumnsAndRows);
      columnsAndRowsRef.current.value = "";
    }
  };

  const onPause = (event: KeyboardEvent): void => {
    if (event.key === " ") {
      setPause(!isPause);
    }
  };

  const handlePause = (): void => {
    setPause(!isPause);
    handleSubmit();
  };

  useEffect(() => {
    document.addEventListener("keydown", onPause);
    return (): void => document.removeEventListener("keydown", onPause);
  }, [isPause]);

  return (
    <StyledOptions>
      <StyledForm onSubmit={handleSubmit}>
        <div>
          <input type="number" ref={timeRef} placeholder="Time in miliseconds" readOnly={!isPause} onChange={onChangeTime} defaultValue={DEFAULT_TIME} min={MIN_TIME} max={MAX_TIME} />
        </div>
        <PlayPause pause={isPause} onClick={(): void => handlePause()} />
        <input
          ref={columnsAndRowsRef}
          placeholder="set number of columns and rows"
          disabled={!isPause}
          onChange={onChangeColumnsAndRows} />
      </StyledForm>
      <StyledTipsAndRulesContainer>
        <button>Tips</button>
        <button>Rules</button>
      </StyledTipsAndRulesContainer>
    </StyledOptions>
  );
};

export default Options;
