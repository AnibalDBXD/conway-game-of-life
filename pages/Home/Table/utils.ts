import Cell from "./Cell";
import { ISearchActiveCellPredicate, ICoordinates, IChangeCells } from "./types";

export const NumberOfCellsX = 40; // Variable
export const NumberOfCellsY = 30; // Variable
export const CellSize = 25;

export const SearchActiveCellPredicate = ({ activeX, activeY, currentX, currentY}: ISearchActiveCellPredicate): boolean => activeX === currentX && activeY === currentY;

export const Rows = Array(NumberOfCellsX).fill({ Component: Cell, isActive: false });

export const Columns = Array(NumberOfCellsY).fill(Rows);

export const getNeighbour = ({ X, Y }: ICoordinates): ICoordinates[] => {
  return [
    //Left Neighnour
    {X: X-1, Y: Y},
    //Right Neighnour
    {X: X+1, Y: Y},
    //Downstairs Neighnour
    {X: X, Y: Y-1},
    //Upstairs Neighbor
    {X: X, Y: Y+1},
    //Neighbor below and left
    {X: X-1, Y: Y-1},
    //Upstairs neighbor and left
    {X: X-1, Y: Y+1},
    //Upstairs neighbor and right
    {X: X+1, Y: Y+1},
    //Neighbor below and right
    {X: X+1, Y: Y-1},
  ];
};

export const searchActiveCell = ({ X: currentX, Y: currentY, activeCells}: IChangeCells): boolean =>
  activeCells.some(({ X: activeX, Y: activeY }) =>
    SearchActiveCellPredicate({ activeX, activeY, currentY, currentX }));

export const deleteCell = ({ X, Y, activeCells }: IChangeCells): ICoordinates[] => {
  const currentActiveCells = [...activeCells];

  const ActiveCellIndex = currentActiveCells.findIndex(({ X: activeX, Y: activeY }) =>
    SearchActiveCellPredicate({ activeX, activeY, currentY: Y, currentX: X }));

  return currentActiveCells.splice(ActiveCellIndex, 1);
};

export const createCell = ({ X, Y, activeCells }: IChangeCells): ICoordinates[] => [...activeCells, {X, Y}];