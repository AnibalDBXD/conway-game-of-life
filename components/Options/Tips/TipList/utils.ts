import { DEFAULT_VALUES } from '../../../../context/Game/utils';
import { createColumnsAndRows } from '../../../Table/utils';

export const squarePattern = (): boolean[][] => {
  const newGame = createColumnsAndRows(DEFAULT_VALUES.COLUMUNS_AND_ROWS);
  const Y = Math.round(newGame.length / 2);
  const X = Math.round(newGame[Y].length / 2);

  newGame[Y][X] = true;
  newGame[Y + 1][X] = true;
  newGame[Y][X + 1] = true;
  newGame[Y + 1][X + 1] = true;
  return newGame;
};

export const boatPattern = (): boolean[][] => {
  const newGame = createColumnsAndRows(DEFAULT_VALUES.COLUMUNS_AND_ROWS);
  const Y = Math.round(newGame.length / 2);
  const X = Math.round(newGame[Y].length / 2);

  newGame[Y + 1][X] = true;
  newGame[Y - 1][X - 1] = true;
  newGame[Y][X - 1] = true;
  newGame[Y][X + 1] = true;
  newGame[Y - 1][X] = true;
  return newGame;
};

export const gliderPattern = (): boolean[][] => {
  const newGame = createColumnsAndRows(DEFAULT_VALUES.COLUMUNS_AND_ROWS);
  const Y = Math.round(newGame.length / 2);
  const X = Math.round(newGame[Y].length / 2);

  newGame[Y - 1][X] = true;
  newGame[Y][X - 1] = true;
  newGame[Y + 1][X] = true;
  newGame[Y - 1][X - 1] = true;
  newGame[Y - 1][X + 1] = true;
  return newGame;
};

export const spaceShipPattern = (): boolean[][] => {
  const newGame = createColumnsAndRows(DEFAULT_VALUES.COLUMUNS_AND_ROWS);
  const Y = Math.round(newGame.length / 2);
  const X = Math.round(newGame[Y].length / 2);

  newGame[Y - 1][X] = true;
  newGame[Y][X - 1] = true;
  newGame[Y + 1][X - 1] = true;
  newGame[Y + 2][X - 1] = true;
  newGame[Y + 2][X] = true;
  newGame[Y + 2][X] = true;
  newGame[Y + 2][X + 1] = true;
  newGame[Y + 2][X + 2] = true;
  newGame[Y + 1][X + 3] = true;
  newGame[Y - 1][X + 3] = true;
  return newGame;
};

export const RpentominoPattern = (): boolean[][] => {
  const newGame = createColumnsAndRows(DEFAULT_VALUES.COLUMUNS_AND_ROWS);
  const Y = Math.round(newGame.length / 2);
  const X = Math.round(newGame[Y].length / 2);

  newGame[Y][X] = true;
  newGame[Y + 1][X] = true;
  newGame[Y - 1][X] = true;
  newGame[Y][X - 1] = true;
  newGame[Y - 1][X + 1] = true;
  return newGame;
};
