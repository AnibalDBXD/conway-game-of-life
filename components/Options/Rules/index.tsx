import Image from 'next/image'

import Button from '../../Button'
import Popup from '../../Popup'
import { IMAGE_SIZE } from '../utils'
import { ImageContainer, StyledFooter } from './styles'

const Rules = (): JSX.Element => {
  return (
    <Popup title="Rules" trigger={<Button>Rules</Button>}>
      <>
        <p>
          This is the Game of Life, it consists on an infinite, two-dimensional{' '}
          <strong>orthogonal</strong> grid of square cells, each of which is in
          one of two possible states, live or dead
        </p>
        <p>
          Every cell interacts with its eight <strong>neighbours</strong>, which
          are the cells that are horizontally, vertically, or diagonally
          adjacent.
        </p>
        <ImageContainer>
          <Image
            src="/img/cell.png"
            alt="Cell without neighbours"
            width={IMAGE_SIZE}
            height={IMAGE_SIZE}
          />
        </ImageContainer>
        <p>At each step in time, the following transitions occur:</p>
        <ul>
          <li>
            Any live cell with fewer than <strong>two</strong> live neighbours{' '}
            <strong>dies</strong>, as if by <strong>underpopulation</strong>.
          </li>

          <li>
            Any live cell with <strong>two or three</strong> live neighbours
            lives on to the next generation.
          </li>

          <li>
            Any live cell with more than <strong>three</strong> live neighbours
            dies, as if by <strong>overpopulation</strong>.
          </li>

          <li>
            Any dead cell with exactly <strong>three</strong> live neighbours
            becomes a live cell, as if by <strong>reproduction</strong>.
          </li>
        </ul>
        <p>
          These rules, which compare the behavior of the automaton to real life,
          can be condensed into the following:
        </p>
        <ul>
          <li>
            Any live cell with <strong>two or three</strong> live neighbours
            survives.
          </li>

          <li>
            Any dead cell with <strong>three</strong> live neighbours becomes a
            live cell.
          </li>

          <li>
            All other live cells die in the next generation. Similarly, all
            other dead cells stay dead.
          </li>
        </ul>

        <StyledFooter
          target="_blank"
          href="https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life"
          rel="noreferrer"
        >
          <small>More information</small>
        </StyledFooter>
      </>
    </Popup>
  )
}

export default Rules
