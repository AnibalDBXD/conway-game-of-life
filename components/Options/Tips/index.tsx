import Button from '../../Button';
import Popup from '../../Popup';
import { StyledSmallTip } from './styles';
import TipList from './TipList';

const Tips = (): JSX.Element => {
  return (
    <Popup title="Tips" trigger={<Button>Tips</Button>}>
      <TipList />
      <StyledSmallTip>
        You can pause the game by pressing the space bar!
      </StyledSmallTip>
    </Popup>
  );
};

export default Tips;
