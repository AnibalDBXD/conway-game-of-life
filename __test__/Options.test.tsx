import React from 'react'
import { act, fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { Primary } from '../stories/Options.stories'
import 'jest-styled-components'
import { DEFAULT_VALUES } from '../context/Game/utils'

describe('<Options />', () => {
  test('Render', () => {
    const { asFragment } = render(<Primary />)
    expect(asFragment()).toMatchSnapshot()
  })
  test("input can't change if pause is true", () => {
    const { getByDisplayValue, getByTestId } = render(<Primary />)
    const timeInput = getByDisplayValue(DEFAULT_VALUES.TIME)
    const columsnAndRowsInput = getByDisplayValue(
      DEFAULT_VALUES.COLUMUNS_AND_ROWS
    )
    const playPlauseButton = getByTestId('PlayPauseButton')

    act(() => {
      fireEvent.click(playPlauseButton)
      const newTimeValue = 800
      fireEvent.change(timeInput, {
        target: {
          value: newTimeValue,
        },
      })
      const newColumnsAndRows = 30
      fireEvent.change(columsnAndRowsInput, {
        target: {
          value: newColumnsAndRows,
        },
      })
      expect(timeInput).toHaveValue(String(DEFAULT_VALUES.TIME))
      expect(columsnAndRowsInput).toHaveValue(
        String(DEFAULT_VALUES.COLUMUNS_AND_ROWS)
      )
    })
  })
})
