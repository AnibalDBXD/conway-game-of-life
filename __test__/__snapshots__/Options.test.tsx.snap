// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Options /> Render 1`] = `
<DocumentFragment>
  .c0 {
  background: linear-gradient(180deg,#313131 7%,#252525 19%);
  height: 20vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c6 {
  border: 0;
  background: transparent;
  box-sizing: border-box;
  width: 0;
  height: 74px;
  border-color: transparent transparent transparent #FFFFFF;
  -webkit-transition: 100ms all ease;
  transition: 100ms all ease;
  cursor: pointer;
  border-style: solid;
  border-width: 37px 0 37px 60px;
}

.c6:hover {
  border-color: transparent transparent transparent #8E8E8E;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 4px;
}

.c4 {
  color: #FFFFFF;
  font-size: 16px;
  font-weight: bold;
}

.c5 {
  font-size: 24px;
  outline: 0;
  border: 0;
  width: 150px;
  border-radius: 10px;
  padding-left: 10px;
  box-shadow: -1px 10px 41px 1px rgba(0,0,0,0.64);
}

.c8 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  background-color: transparent;
  border: 2px solid #FFFFFF;
  border-radius: 8px;
  color: #FFFFFF;
  font-weight: bold;
  cursor: pointer;
  -webkit-transition: -webkit-transform 0.1s ease;
  -webkit-transition: transform 0.1s ease;
  transition: transform 0.1s ease;
}

.c8:hover {
  border-color: #FFFFFFAE;
  color: #FFFFFFAE;
}

.c8:active {
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="Time (miliseconds)"
          >
            Time (miliseconds)
          </label>
          <input
            class="c5"
            id="Time (miliseconds)"
            max="1000"
            min="100"
            type="range"
            value="200"
          />
        </div>
      </div>
      <button
        class="c6"
        data-testid="PlayPauseButton"
      />
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="Columns and Rows"
          >
            Columns and Rows
          </label>
          <input
            class="c5"
            id="Columns and Rows"
            max="40"
            min="8"
            type="range"
            value="16"
          />
        </div>
      </div>
    </div>
    <div
      class="c7"
    >
      <button
        aria-describedby="popup-1"
        class="c8"
      >
        Tips
      </button>
      <button
        class="c8"
      >
        Apply changes
      </button>
      <button
        class="c8"
      >
        Reset
      </button>
      <button
        aria-describedby="popup-2"
        class="c8"
      >
        Rules
      </button>
    </div>
  </div>
</DocumentFragment>
`;
